Support nested model

iche033 opened the pull request:
depends on ignitionrobotics/ign-physics#86 and osrf/sdformat#316

This PR updates the physics system to support parsing and constructing nested models from Model SDF DOM (see osrf/sdformat#316). 

Since the physics system relies heavily on canonical links for setting and also getting pose information of models in the physics engine, some changes were made to support the case where the canonical link is inside a nested model, e.g.:

```
<model name="top_level_model">
  <model name="nested_model">
      <link name="nested_link"> ... </link>
      ...
  </model>
</model> 
```

In the example above, the top level model has one nested model and no child links. In this case, the canonical link would be set to the link inside the nested model.  A couple of helper functions (`TopLevelModel` and `RelativePose`) were added to help compute the transform from the top level model to the nested link in order to correctly set and get model poses. 
Note that I considered using ign-physics's `ModelFrameSemantics` feature to compute this transform data but the feature is currently not part of the minimum feature list, so it's probably not a good idea to depend on an optional physics feature when computing pose updates for a core part of the physics system.

Here's an example nested model world:

```
ign gazebo -v 4 --physics-engine ignition-physics-tpe-plugin nested_model.sdf
```

luca-della-vedova commented:
CC @Kevinskwk to test with our nested model simulation environment.

luca-della-vedova commented on src/systems/physics/Physics.cc:
```diff
@@ -1365,8 +1390,12 @@ void PhysicsPrivate::UpdatePhysics(EntityComponentManager &_ecm)
         auto canonicalPoseComp =
             _ecm.Component<components::Pose>(linkEntityIt->second);
```
From a quick check this variable seems unused now, can it be removed?

iche033 replied:
removed. 7740d5cc86ab37ed863e13e377bb551367c2367b

luca-della-vedova commented on src/systems/physics/Physics.cc:
```diff
@@ -1549,43 +1641,50 @@ void PhysicsPrivate::UpdateSim(EntityComponentManager &_ecm)
-            *(mutableParentPose) = components::Pose(_pose->Data().Inverse() +
-                                           math::eigen3::convert(worldPose));
+            // This is the canonical link, update the top level model.
+            // The pose this link w.r.t its top level model never changes
```
Thought might as well fix the typo here (pose 'of' this link)

iche033 replied:
fixed. 7740d5cc86ab37ed863e13e377bb551367c2367b

luca-della-vedova commented:
I don't know if this is slightly outside the scope. From what I understand the support was only introduced for TPE (which is OK), however when running a world with nested models with DART the behavior changed from printing an error about nested models not being supported (before the PR), to a segmentation fault with the attached backtrace.
Is this intended? Should there be a feature in the physics engine to tell whether nested models are supported or not and avoid trying to spawn them if it is not supported?

```
Thread 1 "ruby" received signal SIGSEGV, Segmentation fault.
0x00007fffbd095530 in dart::simulation::World::getNumSkeletons() const () from /usr/lib/x86_64-linux-gnu/libdart.so.6.10
(gdb) bt
#0  0x00007fffbd095530 in dart::simulation::World::getNumSkeletons() const () from /usr/lib/x86_64-linux-gnu/libdart.so.6.10
#1  0x00007fffbd6d5c46 in ignition::physics::dartsim::Base::AddModel (_worldID=26, _info=..., this=0x55555e7ec908)
    at /usr/include/c++/8/bits/shared_ptr_base.h:1018
#2  ignition::physics::dartsim::SDFFeatures::ConstructSdfModel(ignition::physics::Identity const&, sdf::v9::Model const&) ()
    at /home/luca/ws_citadel/src/ign-physics/dartsim/src/SDFFeatures.cc:359
#3  0x00007fffd8410a5c in ignition::physics::sdf::ConstructSdfModel::Model<ignition::physics::FeaturePolicy<double, 3ul>, ignition::gazebo::v4::systems::PhysicsPrivate::MinimumFeatureList>::ConstructModel (_model=..., this=0x7fffffffb208)
    at /home/luca/ws_citadel/install/include/ignition/plugin1/ignition/plugin/detail/SpecializedPlugin.hh:236
#4  ignition::gazebo::v4::systems::PhysicsPrivate::CreatePhysicsEntities(ignition::gazebo::v4::EntityComponentManager const&)::{lambda(unsigned long const&, ignition::gazebo::v4::components::Component<std::add_lvalue_reference<void>, ignition::gazebo::v4::components::ModelTag, ignition::gazebo::v4::serializers::DefaultSerializer<std::add_lvalue_reference<void> > > const*, ignition::gazebo::v4::components::Component<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, ignition::gazebo::v4::components::NameTag, ignition::gazebo::v4::serializers::StringSerializer> const*, ignition::gazebo::v4::components::Component<ignition::math::v6::Pose3<double>, ignition::gazebo::v4::components::PoseTag, ignition::gazebo::v4::serializers::DefaultSerializer<ignition::math::v6::Pose3<double> > > const*, ignition::gazebo::v4::components::Component<unsigned long, ignition::gazebo::v4::components::ParentEntityTag, ignition::gazebo::v4::serializers::DefaultSerializer<unsigned long> > const*)#2}::operator()(unsigned long const&, ignition::gazebo::v4::components::Component<std::add_lvalue_reference<void>, ignition::gazebo::v4::components::ModelTag, ignition::gazebo::v4::serializers::DefaultSerializer<std::add_lvalue_reference<void> > > const*, ignition::gazebo::v4::components::Component<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, ignition::gazebo::v4::components::NameTag, ignition::gazebo::v4::serializers::StringSerializer> const*, ignition::gazebo::v4::components::Component<ignition::math::v6::Pose3<double>, ignition::gazebo::v4::components::PoseTag, ignition::gazebo::v4::serializers::DefaultSerializer<ignition::math::v6::Pose3<double> > > const*, ignition::gazebo::v4::components::Component<unsigned long, ignition::gazebo::v4::components::ParentEntityTag, ignition::gazebo::v4::serializers::DefaultSerializer<unsigned long> > const*) const [clone .isra.2050] () at /home/luca/ws_citadel/src/ign-gazebo/src/systems/physics/Physics.cc:712
#5  0x00007fffd84594cd in std::function<bool (unsigned long const&, ignition::gazebo::v4::components::Component<std::add_lvalue_reference<void>, ignition::gazebo::v4::components::ModelTag, ignition::gazebo::v4::serializers::DefaultSerializer<std::add_lvalue_reference<void> > > const*, ignition::gazebo::v4::components::Component<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, ignition::gazebo::v4::components::NameTag, ignition::gazebo::v4::serializers::StringSerializer> const*, ignition::gazebo::v4::components::Component<ignition::math::v6::Pose3<double>, ignition::gazebo::v4::components::PoseTag, ignition::gazebo::v4::serializers::DefaultSerializer<ignition::math::v6::Pose3<double> > > const*, ignition::gazebo::v4::components::Component<unsigned long, ignition::gazebo::v4::components::ParentEntityTag, ignition::gazebo::v4::serializers::DefaultSerializer<unsigned long> > const*)>::operator()(unsigned long const&, ignition::gazebo::v4::components::Component<std::add_lvalue_reference<void>, ignition::gazebo::v4::components::ModelTag, ignition::gazebo::v4::serializers::DefaultSerializer<std::add_lvalue_reference<void> > > const*, ignition::gazebo::v4::components::Component<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, ignition::gazebo::v4::components::NameTag, ignition::gazebo::v4::serializers::StringSerializer> const*, ignition::gazebo::v4::components::Component<ignition::math::v6::Pose3<double>, ignition::gazebo::v4::components::PoseTag, ignition::gazebo::v4::serializers::DefaultSerializer<ignition::math::v6::Pose3<double> > > const*, ignition::gazebo::v4::components::Component<unsigned long, ignition::gazebo::v4::components::ParentEntityTag, ignition::gazebo::v4::serializers::DefaultSerializer<unsigned long> > const*) const (__args#4=<optimized out>, __args#3=<optimized out>, 
    __args#2=<optimized out>, __args#1=<optimized out>, __args#0=@0x7fffffffb380: 220, this=0x7fffffffb410)
    at /usr/include/c++/8/bits/std_function.h:682
#6  ignition::gazebo::v4::EntityComponentManager::EachNew<ignition::gazebo::v4::components::Component<std::add_lvalue_reference<void>, ignition::gazebo::v4::components::ModelTag, ignition::gazebo::v4::serializers::DefaultSerializer<std::add_lvalue_reference<void> > >, ignition::gazebo::v4::components::Component<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, ignition::gazebo::v4::components::NameTag, ignition::gazebo::v4::serializers::StringSerializer>, ignition::gazebo::v4::components::Component<ignition::math::v6::Pose3<double>, ignition::gazebo::v4::components::PoseTag, ignition::gazebo::v4::serializers::DefaultSerializer<ignition::math::v6::Pose3<double> > >, ignition::gazebo::v4::components::Component<unsigned long, ignition::gazebo::v4::components::ParentEntityTag, ignition::gazebo::v4::serializers::DefaultSerializer<unsigned long> > >(ignition::gazebo::v4::EntityComponentManager::identity<std::function<bool (unsigned long const&, ignition::gazebo::v4::components::Component<std::add_lvalue_reference<void>, ignition::gazebo::v4::components::ModelTag, ignition::gazebo::v4::serializers::DefaultSerializer<std::add_lvalue_reference<void> > > const*, ignition::gazebo::v4::components::Component<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, ignition::gazebo::v4::components::NameTag, ignition::gazebo::v4::serializers::StringSerializer> const*, ignition::gazebo::v4::components::Component<ignition::math::v6::Pose3<double>, ignition::gazebo::v4::components::PoseTag, ignition::gazebo::v4::serializers::DefaultSerializer<ignition::math::v6::Pose3<double> > > const*, ignition::gazebo::v4::components::Component<unsigned long, ignition::gazebo::v4::components::ParentEntityTag, ignition::gazebo::v4::serializers::DefaultSerializer<unsigned long> > const*)> >::type) const (this=this@entry=0x555555c2c620, _f=...)
    at /home/luca/ws_citadel/src/ign-gazebo/include/ignition/gazebo/detail/EntityComponentManager.hh:348
#7  0x00007fffd8414f75 in ignition::gazebo::v4::systems::PhysicsPrivate::CreatePhysicsEntities(ignition::gazebo::v4::EntityComponentManager const&) () at /usr/include/c++/8/new:169
#8  0x00007fffd8415f38 in ignition::gazebo::v4::systems::Physics::Update(ignition::gazebo::v4::UpdateInfo const&, ignition::gazebo::v4::EntityComponentManager&) () at /home/luca/ws_citadel/src/ign-gazebo/src/systems/physics/Physics.cc:619
#9  0x00007ffff4a682d2 in ignition::gazebo::v4::SimulationRunner::UpdateSystems (this=this@entry=0x555555c2c520)
    at /home/luca/ws_citadel/src/ign-gazebo/src/SimulationRunner.cc:459
#10 0x00007ffff4a6dbd7 in ignition::gazebo::v4::SimulationRunner::Step (this=this@entry=0x555555c2c520, _info=...)
    at /home/luca/ws_citadel/src/ign-gazebo/src/SimulationRunner.cc:696
#11 0x00007ffff4a6e29b in ignition::gazebo::v4::SimulationRunner::Run(unsigned long) ()
    at /home/luca/ws_citadel/src/ign-gazebo/src/SimulationRunner.cc:669
#12 0x00007ffff4a62915 in ignition::gazebo::v4::ServerPrivate::Run(unsigned long, std::optional<std::condition_variable*>) ()
    at /usr/include/c++/8/bits/unique_ptr.h:345
#13 0x00007ffff4a5db11 in ignition::gazebo::v4::Server::Run(bool, unsigned long, bool) () at /usr/include/c++/8/optional:969
#14 0x00007ffff5184918 in runServer () at /home/luca/ws_citadel/src/ign-gazebo/src/ign.cc:267
#15 0x00007ffff5592dae in ffi_call_unix64 () from /usr/lib/x86_64-linux-gnu/libffi.so.6
#16 0x00007ffff559271f in ffi_call () from /usr/lib/x86_64-linux-gnu/libffi.so.6
#17 0x00007ffff57998f8 in ?? () from /usr/lib/x86_64-linux-gnu/ruby/2.5.0/fiddle.so
#18 0x00007ffff7aa3cab in ?? () from /usr/lib/x86_64-linux-gnu/libruby-2.5.so.2.5
#19 0x00007ffff579968d in ?? () from /usr/lib/x86_64-linux-gnu/ruby/2.5.0/fiddle.so
#20 0x00007ffff7aca3c9 in ?? () from /usr/lib/x86_64-linux-gnu/libruby-2.5.so.2.5
#21 0x00007ffff7ad88f3 in ?? () from /usr/lib/x86_64-linux-gnu/libruby-2.5.so.2.5
#22 0x00007ffff7aced85 in ?? () from /usr/lib/x86_64-linux-gnu/libruby-2.5.so.2.5
#23 0x00007ffff7ad4b64 in ?? () from /usr/lib/x86_64-linux-gnu/libruby-2.5.so.2.5
#24 0x00007ffff79af0c4 in ?? () from /usr/lib/x86_64-linux-gnu/libruby-2.5.so.2.5
#25 0x00007ffff79b0f4d in ruby_exec_node () from /usr/lib/x86_64-linux-gnu/libruby-2.5.so.2.5
#26 0x00007ffff79b342e in ruby_run_node () from /usr/lib/x86_64-linux-gnu/libruby-2.5.so.2.5
#27 0x00005555555548cb in ?? ()
#28 0x00007ffff7554b97 in __libc_start_main (main=0x555555554880, argc=5, argv=0x7fffffffc7a8, init=<optimized out>, fini=<optimized out>, 
    rtld_fini=<optimized out>, stack_end=0x7fffffffc798) at ../csu/libc-start.c:310
#29 0x00005555555548fa in _start ()

```

iche033 commented:
I added a check in 7740d5cc86ab37ed863e13e377bb551367c2367b that to make sure the constructed model is not null and warned the user if it is. Note you'll also need to update the `tpe_nested_model` branch in ign-physics. This may be sufficient for now to tell whether or not the physics engine supports nested model. I think there are also plans to add nested model support to dartsim plugin in ign-physics and we could consider adding a nested model feature.

luca-della-vedova commented:
> I added a check in [7740d5c](https://github.com/ignitionrobotics/ign-gazebo/commit/7740d5cc86ab37ed863e13e377bb551367c2367b) that to make sure the constructed model is not null and warned the user if it is. Note you'll also need to update the `tpe_nested_model` branch in ign-physics. This may be sufficient for now to tell whether or not the physics engine supports nested model. I think there are also plans to add nested model support to dartsim plugin in ign-physics and we could consider adding a nested model feature.

Tried it out and indeed the crash is fixed, thanks! I'll try to get a full tpe sim up and running but that is quite a daunting task so don't keep it as a blocker to get this merged :sweat_smile:

claireyywang commented on test/integration/physics_system.cc:
```diff
@@ -875,3 +875,110 @@ TEST_F(PhysicsSystemFixture, GetBoundingBox)
+  server.AddSystem(testSystem.systemPtr);
+  server.Run(true, 1, false);
+
+  EXPECT_EQ(2u, postUpModelPoses.size());
+  EXPECT_EQ(2u, postUpLinkPoses.size());
+  EXPECT_EQ(4u, parents.size());
```
found some test failures here 
```
/home/claire/citadel_ws/src/ign-gazebo/test/integration/physics_system.cc:949: Failure
Value of: postUpModelPoses.size()
  Actual: 1
Expected: 2u
Which is: 2
/home/claire/citadel_ws/src/ign-gazebo/test/integration/physics_system.cc:950: Failure
Value of: postUpLinkPoses.size()
  Actual: 1
Expected: 2u
Which is: 2
/home/claire/citadel_ws/src/ign-gazebo/test/integration/physics_system.cc:951: Failure
Value of: parents.size()
  Actual: 2
Expected: 4u
Which is: 4
/home/claire/citadel_ws/src/ign-gazebo/test/integration/physics_system.cc:958: Failure
Expected: (postUpModelPoses.end()) != (modelIt), actual: 8-byte object <00-00 00-00 00-00 00-00> vs 8-byte object <00-00 00-00 00-00 00-00>

```

iche033 replied:
hmm weird I'm not able to reproduce this. Will wait for another jenkins results and see if the builds catch this as well.

claireyywang replied:
I updated all repo and reran the tests and the failure went away!

iche033 commented on src/systems/physics/Physics.cc:
```diff
@@ -1549,43 +1646,50 @@ void PhysicsPrivate::UpdateSim(EntityComponentManager &_ecm)
+               _ecm.Component<components::Pose>(topLevelModel);
+            *(mutableModelPose) = components::Pose(
+                linkPoseFromTopLevelModel.Inverse() +
+                math::eigen3::convert(worldPose));
```
updated. def640a

iche033 commented:
>  Nevertheless, my preference would be to create the new feature and document that it is not complete.

ok I added ConstructSdfNestedModel feature in ignitionrobotics/ign-physics#86 and updated the physics system to use this optional feature in 57c2d1d7c24eded658d234ebb31ed7136ec9138e

> If we create the new feature, we can formally limit the ConstructSdfModel feature to work with non-nested models.

I have not added this limitation yet. To confirm, this means we need to add `ConstructSdfNestedModel` feature to both `WorldPtrType` and `ModelPtrType`? Currently I use `world->ConstructModel` and `model->ConstructNestedModel`, both of them can consume an `sdf::Model` object with nested models. If we limit `ConstructSdfModel` feature to non-nested models, we would need to do an additional check on the input `sdf::Model` to see if it has child models before determining which Construct feature to use, correct?

azeey commented:
> To confirm, this means we need to add ConstructSdfNestedModel feature to both WorldPtrType and ModelPtrType? Currently I use world->ConstructModel and model->ConstructNestedModel, both of them can consume an sdf::Model object with nested models. If we limit ConstructSdfModel feature to non-nested models, we would need to do an additional check on the input sdf::Model to see if it has child models before determining which Construct feature to use, correct?

Yes, this is what I was thinking. @scpeters, does this sound reasonable?

azeey commented on src/systems/physics/Physics.cc:
```diff
@@ -1549,43 +1679,50 @@ void PhysicsPrivate::UpdateSim(EntityComponentManager &_ecm)
-              _ecm.Component<components::Pose>(_parent->Data());
-            *(mutableParentPose) = components::Pose(_pose->Data().Inverse() +
-                                           math::eigen3::convert(worldPose));
+            // This is the canonical link, update the top level model.
```
This might be problematic because nested models have their own canonical links in addition to the canonical link of the containing model. While iterating through the links, you might have multiple canonical links that all have the same top level model. I'm not sure what would be the best way to avoid this. It would be nice if we could tell which canonical link belongs to which model in the hierarchy. This would make it possible to change the pose of nested models using the transform tool, something that isn't available in [gazebo-classic](https://answers.gazebosim.org/question/25506/sdf-worldmodel-composition-across-files-is-impossible-groupingempty-containers-of-models/). But right now the `components::CanonicalLink` component doesn't give us enough information to make that association. Maybe we can add a pointer to the model entity in the component, but that will have to wait for a future version because it breaks ABI. 

For now, the only solution I can think of is to ensure that only one canonical link per model hierarchy exists by construction in SdfEntityCreator.

iche033 replied:
sounds good. I added a helper function `SdfEntityCreator::CreateEntities(sdf::Model _model, bool _createCanonicalLink)` function that recusrively creates the model while keeping track of whether a canonical link has been created or not in a model tree.

iche033 commented on src/systems/physics/Physics.cc:
```diff
@@ -1362,11 +1428,12 @@ void PhysicsPrivate::UpdatePhysics(EntityComponentManager &_ecm)
-            _ecm.Component<components::Pose>(linkEntityIt->second);
+        // set world pose of canonical link in freegroup
+        math::Pose3d linkPose =
+            this->RelativePose(_entity, linkEntityIt->second, _ecm);
```
yep added. f585388

iche033 commented on src/systems/physics/Physics.cc:
```diff
@@ -1536,9 +1663,12 @@ void PhysicsPrivate::UpdateSim(EntityComponentManager &_ecm)
+        // If model is static, don't process pose changes as periodic
         const auto *staticComp =
-          _ecm.Component<components::Static>(_parent->Data());
+          _ecm.Component<components::Static>(topLevelModel);
```
ok I see, it seems like the proposal allows the child nested models to be static while the top level model can be non-static but not the other way around? So: 

* if top level model is static, all nested models in the tree will be static. 
* if top level model is non-static, nested models can be static or non-static

implemented in c070152

iche033 commented:
> Yes, this is what I was thinking. @scpeters, does this sound reasonable?

I added a new `ContructedSdfNestedModel` to `WorldPtrType` in  ignitionrobotics/ign-physics#86 and updated the physics system to make use of this new optional feature. c070152

scpeters commented:
> > To confirm, this means we need to add ConstructSdfNestedModel feature to both WorldPtrType and ModelPtrType? Currently I use world->ConstructModel and model->ConstructNestedModel, both of them can consume an sdf::Model object with nested models. If we limit ConstructSdfModel feature to non-nested models, we would need to do an additional check on the input sdf::Model to see if it has child models before determining which Construct feature to use, correct?
> 
> Yes, this is what I was thinking. @scpeters, does this sound reasonable?

yes, sounds reasonable to me too

azeey commented on src/systems/physics/Physics.cc:
```diff
@@ -1362,11 +1469,14 @@ void PhysicsPrivate::UpdatePhysics(EntityComponentManager &_ecm)
 
-        auto canonicalPoseComp =
-            _ecm.Component<components::Pose>(linkEntityIt->second);
+        // set world pose of canonical link in freegroup
```
Here's an issue when setting the pose of a nested model. Maybe we should allow `WorldPoseCmd` only on top level models? This probably applies to other `*Cmd` components as well.
![Peek 2020-08-12 00-01](https://user-images.githubusercontent.com/206116/89977084-e1e9f000-dc2f-11ea-948e-cd55c04ac5e2.gif)

iche033 replied:
ok I  disabled gui for setting pose cmds in 6db69f494d315b09d26eabdf9edf0ca314c4d4c8 and prevented setting pose and vel commands in physics system in 451959056358b2a92041ca5ee2a448b21bc35d33

iche033 commented on include/ignition/gazebo/SdfEntityCreator.hh:
```diff
@@ -155,6 +155,15 @@ namespace ignition
+      /// \param[in] _createCanonicalLink True to create a canonical link
+      /// component and attach to its child link entity
+      /// \return Model entity.
+      private: Entity CreateEntities(const sdf::Model *_model,
```
it calls other functions in the `SdfEntityCreator` class (like `SetParent` and `CreateEntities` for links` so we would have to either create those versions in `SdfEntityCreatePrivate` as well or pass in a pointer to the public class. It gets a little messy so I left it a private function.

iche033 commented on src/SdfEntityCreator.cc:
```diff
@@ -238,18 +250,22 @@ Entity SdfEntityCreator::CreateEntities(const sdf::Model *_model)
   // the parent frame until we get frames working.
 
   // Links
+  bool createdCanonicalLink = false;
```
changed variable name. 6db69f494d315b09d26eabdf9edf0ca314c4d4c8

iche033 commented on src/systems/physics/Physics.cc:
```diff
@@ -664,36 +710,97 @@ void PhysicsPrivate::CreatePhysicsEntities(const EntityComponentManager &_ecm)
+            auto parentStaticComp =
+              _ecm.Component<components::Static>(_parent->Data());
+            if (parentStaticComp && parentStaticComp->Data())
+              model.SetStatic(true);
```
ok I think 6db69f494d315b09d26eabdf9edf0ca314c4d4c8 does what you have in mind. The static component is set and overriden when the entity is created.

iche033 commented on src/systems/physics/Physics.cc:
```diff
@@ -664,36 +710,97 @@ void PhysicsPrivate::CreatePhysicsEntities(const EntityComponentManager &_ecm)
+            else
+            {
+              ignerr << "Model: '" << _name->Data() << "' not loaded. "
+                     << "Nested model not supported by physics engine. "
```
updated error msg. 6db69f494d315b09d26eabdf9edf0ca314c4d4c8

iche033 commented on src/systems/physics/Physics.cc:
```diff
@@ -1536,56 +1706,66 @@ void PhysicsPrivate::UpdateSim(EntityComponentManager &_ecm)
+            }
+            math::Pose3d parentWorldPose =
+                this->RelativePose(worldComp->Data(), _parent->Data(), _ecm);
+
```
great thanks!

iche033 commented on src/systems/physics/Physics.cc:
```diff
@@ -1536,56 +1706,66 @@ void PhysicsPrivate::UpdateSim(EntityComponentManager &_ecm)
+
             *_pose = components::Pose(math::eigen3::convert(worldPose) +
-                                      parentPose->Data().Inverse());
+                                      parentWorldPose.Inverse());
```
updated comment. 6db69f494d315b09d26eabdf9edf0ca314c4d4c8

chapulina commented on include/ignition/gazebo/SdfEntityCreator.hh:
```diff
@@ -155,6 +155,18 @@ namespace ignition
+      /// \param[in] _staticParent True if parent is static, false otherwise.
+      /// \return Model entity.
+      private: Entity CreateEntities(const sdf::Model *_model,
+          bool _createCanonicalLink, bool _staticParent);
```
Is static the only option that the nested models will ever override? Maybe consider passing some type that can be extended in the future. A map with key-value pairs? Or maybe it's not worth it...

iche033 replied:
I'm not sure what other params will be added. This function has evolved as I get more info about the nested model proposal. I'm leaning towards leaving as is for now since it's private and we can change it later if we need too.

chapulina commented on src/systems/physics/Physics.cc:
```diff
@@ -664,36 +710,97 @@ void PhysicsPrivate::CreatePhysicsEntities(const EntityComponentManager &_ecm)
+                     << " engine doesn't support feature "
+                     << "[ConstructSdfNestedModelFeature]. "
+                     << "Nested model will be ignored."
+                     << std::endl;
```
If there's a chance this will spam the console, consider using the `static bool informed` pattern that's used in other parts of this file.

iche033 replied:
done. a71a3d2e2e65bab10243987f91ed238a1329f7ab

chapulina commented on src/systems/physics/Physics.cc:
```diff
@@ -190,6 +191,21 @@ class ignition::gazebo::systems::PhysicsPrivate
+  /// \param[in] _entity Input entity
+  /// \param[in] _ecm Constant reference to ECM.
+  /// \return Entity of top level model
+  public: ignition::gazebo::Entity TopLevelModel(const Entity &_entity,
```
I think these functions could be in `Util.hh`, they should be useful to other systems. Then we can also write some unit tests for them.

iche033 replied:
moved to Util class. a71a3d2e2e65bab10243987f91ed238a1329f7ab

chapulina commented on src/systems/physics/Physics.cc:
```diff
@@ -1536,56 +1730,70 @@ void PhysicsPrivate::UpdateSim(EntityComponentManager &_ecm)
           const components::ParentEntity *_parent)->bool
       {
-        // If parent is static, don't process pose changes as periodic
+        // If model is static, don't process pose changes as periodic
```
The entity inside this `Each` call should be a link though, and I believe links shouldn't have static components, right?

iche033 replied:
fixed. a71a3d2e2e65bab10243987f91ed238a1329f7ab

chapulina commented on src/systems/physics/Physics.cc:
```diff
@@ -1536,56 +1730,70 @@ void PhysicsPrivate::UpdateSim(EntityComponentManager &_ecm)
+
+            // Note that if the canonical link is inside a nested model, the
+            // immediate parent model of this link will not have its pose
+            // updated. Nested models have a fixed pose relative to the top
```
> Nested models have a fixed pose relative to the top level model.

I think this is only true for the nested model that has the canonical link, right? Maybe the comment can be updated to avoid confusion?

iche033 replied:
actually this comment is kind of out-of-place as the code here is for mainly about non-canonical links. I've updated the comment in a71a3d2e2e65bab10243987f91ed238a1329f7ab

chapulina commented on test/worlds/nested_model.sdf:
```diff
@@ -0,0 +1,66 @@
+<?xml version="1.0" ?>
+<sdf version="1.6">
```
1.7?

iche033 replied:
updated. a71a3d2e2e65bab10243987f91ed238a1329f7ab

chapulina commented on test/integration/physics_system.cc:
```diff
@@ -875,3 +875,110 @@ TEST_F(PhysicsSystemFixture, GetBoundingBox)
+
+/////////////////////////////////////////////////
+// This tests whether nested models can be loaded correctly
+TEST_F(PhysicsSystemFixture, NestedModel)
```
Would you mind adding a few more tests?

* loading a plugin for a nested model - bonus for loading it both from the SDF file and `ServerConfig` :slightly_smiling_face: 
* saving a world that has a nested model to make sure they're saved properly
* maybe also some test to make sure they can be logged and played back

iche033 replied:
added more tests in a71a3d2e2e65bab10243987f91ed238a1329f7ab and 4a16b5585e9d3967229fd26da3bafeb803b4a86a

chapulina reviewed (commented):
Just took a high level view, have some minor comments.

azeey reviewed (approved):
I'm approving this PR, but I want to note that we need to address the issue of nested model poses at some point. This PR only updates the pose of the top level model and leaves the poses of nested models at their initial state. However, a number of plugins (eg. Breadcrumbs) rely on the model's pose in their computation, which means they will fail if the model they are in is nested inside another model. This is not an issue right now since TPE doesn't support joints and pose and velocity commands are disabled for nested models.

iche033 commented on examples/worlds/nested_model.sdf:
```diff
@@ -0,0 +1,110 @@
+<?xml version="1.0" ?>
+<sdf version="1.6">
```
done. 6b741439ba65dd7cec865b1e31b7c7511c5a55e8

iche033 commented on src/SdfEntityCreator.cc:
```diff
@@ -216,6 +220,41 @@ Entity SdfEntityCreator::CreateEntities(const sdf::Model *_model)
+  auto ent = this->CreateEntities(_model, true, false);
+
+  // Load all model plugins afterwards, so we get scoped name for nested models.
+  for (const auto &[entity, element] : this->dataPtr->newModels)
```
This was mainly done so the all the ancestor models are loaded first before loading the nested model's plugin so that it's able to get the scoped name. As for the order of loading the plugins, I think it maybe fine but not 100% certain. I just checked the behavior in gazebo classic and it's also loading the parent model's plugin first before the children's plugins. So we could give this a try and come back if we run into issues in the future?

azeey replied:
Sounds good to me.

iche033 commented:
ok I think I addressed all comments and ticketed issue #392 about the nested model pose issue.

iche033 commented:
@osrf-jenkins run tests please

scpeters commented:
when finding the sdformat9 version in cmake, I think we should require version 9.3.0

iche033 commented:
bumped sdformat version requirment to 9.3.0 in 44cd092

chapulina commented:
Ouch, this should have been squash-merged. Now all of this PR's commits are scattered around the `ign-gazebo3` branch, which makes it hard to identify what changes came from each PR:

https://github.com/ignitionrobotics/ign-gazebo/compare/ignition-gazebo3_3.3.0...ign-gazebo3

iche033 commented:
ohh oops, sorry. I thought it defaulted to squash and merge. I'll make sure to double check next time.

chapulina commented:
> I thought it defaulted to squash and merge.

Yeah it annoyingly defaults to the last used, which may be "merge commit" or "rebase and merge" in case of ports across branches :confused:
